src/tbx/mod.rs

```diff
@@ -0,0 +1,410 @@
+            s: ptr::null_mut(),
+        };
+        unsafe {
+            while htslib::hts_getline(hts_file, KS_SEP_LINE, &mut buf) >= 0 {
+                if buf.l > 0 && (*buf.s) as i32 == (*tbx).conf.meta_char {
+                    header.push(String::from(ffi::CStr::from_ptr(buf.s).to_str().unwrap()));
+                } else {
+                    break;
+                }
+            }
+        }
+
+        let mut nseq: i32 = 0;
+        let seqs = unsafe { htslib::tbx_seqnames(tbx, &mut nseq) };
+        let mut seqnames = Vec::new();
+        for i in 0..nseq {
+            unsafe {
+                seqnames.push(String::from(
+                    ffi::CStr::from_ptr(*seqs.offset(i as isize))
+                        .to_str()
+                        .unwrap(),
+                ));
+            }
+        }
+        unsafe {
+            libc::free(seqs as (*mut libc::c_void));
+        };
+
+        if tbx.is_null() {
+            Err(TabixReaderError::InvalidIndex)
+        } else {
+            Ok(TabixReader {
+                header,
+                seqnames,
+                hts_file,
+                hts_format,
+                tbx,
+                buf,
+                itr: None,
+                tid: -1,
+                beg: -1,
+                end: -1,
+            })
+        }
+    }
+
+    /// Get sequence ID from sequence name.
+    pub fn seq_name_to_id(&self, name: &str) -> Result<u32, SequenceLookupError> {
+        // TODO: naming?
+        let res = unsafe {
+            htslib::tbx_name2id(
+                self.tbx,
+                ffi::CString::new(name.as_bytes()).unwrap().as_ptr(),
+            )
+        };
+        if res < 0 {
+            Err(SequenceLookupError::Some)
+        } else {
+            Ok(res as u32)
+        }
+    }
+
+    /// Fetch region given by numeric sequence number and 1-based begin and end position.
+    pub fn fetch(&mut self, tid: u32, beg: u32, end: u32) -> Result<(), FetchError> {
+        self.tid = tid as i32;
+        self.beg = beg as i32;
+        self.end = end as i32;
+
+        if let Some(itr) = self.itr {
+            unsafe {
+                htslib::hts_itr_destroy(itr);
+            }
+        }
+        let itr = unsafe {
+            htslib::hts_itr_query(
+                (*self.tbx).idx,
+                tid as i32,
+                beg as i32,
+                end as i32,
+                Some(htslib::tbx_readrec),
+            )
+        };
+        if itr.is_null() {
+            self.itr = None;
+            Err(FetchError::Some)
+        } else {
+            self.itr = Some(itr);
+            Ok(())
+        }
+    }
+}
+
+/// Return whether the two given genomic intervals overlap.
+fn overlap(tid1: i32, begin1: i32, end1: i32, tid2: i32, begin2: i32, end2: i32) -> bool {
+    (tid1 == tid2) && (begin1 < end2) && (begin2 < end1)
+}
+
+impl Read for TabixReader {
+    fn read(&mut self, record: &mut String) -> Result<(), ReadError> {
```

Comment:
I think this should be a `Vec<u8>`, i.e., an ASCII string?

Reply:
OK